pkg/kubelet/kuberuntime/instrumented_services_test.go

```diff
@@ -29,10 +30,17 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
 )
 
+var registerMetrics sync.Once
+
 func TestRecordOperation(t *testing.T) {
-	legacyregistry.MustRegister(metrics.RuntimeOperations)
-	legacyregistry.MustRegister(metrics.RuntimeOperationsDuration)
-	legacyregistry.MustRegister(metrics.RuntimeOperationsErrors)
+
+	registerMetrics.Do(func() {
+		legacyregistry.MustRegister(metrics.RuntimeOperations)
+		legacyregistry.MustRegister(metrics.RuntimeOperationsDuration)
+		legacyregistry.MustRegister(metrics.RuntimeOperationsErrors)
+	},
+	)
+	defer legacyregistry.Reset()
```

Comment:
This looks problematic to me, it assumes that there are no concurrent tests using the legacyregistry.

Reply:
how can we deal with it then?

Reply:
/cc @logicalhan

Reply:
Let's first be explicit about whether there is a problem.
Are there test scenarios that have concurrent test func invocations in one process?

If so then the implication is that any given test func X can not assume that a registered metric is being affected only by the current invocation of X.

Reply:
Hi Mike, 
these tests are executed one by one. These tests don't overlap.

I printed begin time and end time for each test in 
https://gist.github.com/CatherineF-dev/7557f0abdabe5e13ba1b37c7fa15523c

Reply:
This problem only exists because we're using the legacyregistry in this test, which is a global. If we instantiated a new registry per test run and registered the metrics to this registry, then we would not have potential collisions across test runs. You would also be free to reset the registry since other test runs would be instantiating their own registry in their runs.

Reply:
I see that in https://gist.github.com/CatherineF-dev/7557f0abdabe5e13ba1b37c7fa15523c the test func invocations were serialized.  My question is, can a test author rely on the test func invocations always being serialized?

Reply:
can we use a new registry here and keep the test behavior?
can you share an example?

Reply:
The sample_and_watermark_test.go updated in https://github.com/kubernetes/kubernetes/pull/105886 could make and use a private registry in each test func invocation.

Same for the metrics_test.go in that same PR.

Reply:
I'm a bit busy these, anyone feels that can take over this PR?

Reply:
I can take it. Since #105886 is a similar issue.

Reply:
During tests, I found one advantage for using the global registry (legacyregistry). It could help detecting duplicated metrics name. 

---------
pkg1:  metrics_a
pkg2: metrics_a
Both pkg1 and pkg2 emit metrics_a. It will be detected by the global registry, which is pretty useful.

Reply:
Duplicate metric names from registrations from non-test code will be detected by other unit, integration, and end-to-end tests.